boofuzz/fuzz_logger_db.py

```diff
@@ -67,49 +71,60 @@ def get_test_case_data(self, index):
-                                [self._current_test_case_index, 'error', description, b'', helpers.get_time_stamp()])
-        self._database_connection.commit()
+        self._queue.append(["INSERT INTO steps VALUES(?, ?, ?, ?, ?);\n",
+            self._current_test_case_index, 'error', description, b'', helpers.get_time_stamp()])
+        self._fail_detected = True
+        self.write_log()
 
     def log_recv(self, data):
-        self._db_cursor.execute('''INSERT INTO steps VALUES(?, ?, ?, ?, ?)''',
-                                [self._current_test_case_index, 'receive', u'', buffer(data), helpers.get_time_stamp()])
-        self._database_connection.commit()
+        self._queue.append(["INSERT INTO steps VALUES(?, ?, ?, ?, ?);\n",
+            self._current_test_case_index, 'receive', u'', buffer(data), helpers.get_time_stamp()])
 
     def log_send(self, data):
-        self._db_cursor.execute('''INSERT INTO steps VALUES(?, ?, ?, ?, ?)''',
-                                [self._current_test_case_index, 'send', u'', buffer(data), helpers.get_time_stamp()])
-        self._database_connection.commit()
+        self._queue.append(["INSERT INTO steps VALUES(?, ?, ?, ?, ?);\n",
+            self._current_test_case_index, 'send', u'', buffer(data), helpers.get_time_stamp()])
 
     def log_info(self, description):
-        self._db_cursor.execute('''INSERT INTO steps VALUES(?, ?, ?, ?, ?)''',
-                                [self._current_test_case_index, 'info', description, b'', helpers.get_time_stamp()])
-        self._database_connection.commit()
+        self._queue.append(["INSERT INTO steps VALUES(?, ?, ?, ?, ?);\n",
+            self._current_test_case_index, 'info', description, b'', helpers.get_time_stamp()])
 
     def log_fail(self, description=""):
-        self._db_cursor.execute('''INSERT INTO steps VALUES(?, ?, ?, ?, ?)''',
-                                [self._current_test_case_index, 'fail', description, b'', helpers.get_time_stamp()])
-        self._database_connection.commit()
+        self._queue.append(["INSERT INTO steps VALUES(?, ?, ?, ?, ?);\n",
+            self._current_test_case_index, 'fail', description, b'', helpers.get_time_stamp()])
+        self._fail_detected = True
 
     def log_pass(self, description=""):
-        self._db_cursor.execute('''INSERT INTO steps VALUES(?, ?, ?, ?, ?)''',
-                                [self._current_test_case_index, 'pass', description, b'', helpers.get_time_stamp()])
-        self._database_connection.commit()
+        self._queue.append(["INSERT INTO steps VALUES(?, ?, ?, ?, ?);\n",
+            self._current_test_case_index, 'pass', description, b'', helpers.get_time_stamp()])
+
+    def write_log(self, force=False):
+
+        if self._queue_max_len > 0:
+            while (self._current_test_case_index - next(
+                    x for x in self._queue[0] if isinstance(x, (int, long)))) >= self._queue_max_len:
+                self._queue.popleft()
```

Comment:
Sorry I should have commented that in advance.
The general idea was to always keep the log of e.g. 3 full test cases in RAM and only write them to disk if an error/fail occurs or fuzzing is complete. By setting deque's maxlen we could only limit the max number of single log lines. As a testcase doesn't always output the same amount of lines, this would end up in logging half test cases.
Another way that this could possibly be done would be with a 2D list inside the deque. That way every test case could get it's own row and as many columns as needed for the actual log messages. But I think this would make appending to the queue more complicated.

Now to your initial question about what that statement does:

1. You are right about the downsizing part. It keeps removing log lines until the max number of test cases is reached.
2. It searches for the first `int` in the oldest queue list, which should be the test case index of that log line. The problem I was facing here was, that the index field is not always on the same positon of the list. For `open_test_case()` it's on the second position, for all others it's on the first.
I didn't want to change the log format or the `get_test_case_data` functions so I went with this solution instead, which is clearly not best practice.

Reply:
@SR4ven That clears up a lot, thank you.